File: slp/backend-dotnet/Middlewares/MetricsMiddleware.cs
using backend_dotnet.Features.Metrics;
using System.Diagnostics;

namespace backend_dotnet.Middlewares;

public class MetricsMiddleware
{
    private readonly RequestDelegate _next;

    // Logger injected via constructor (singleton-safe)
    public MetricsMiddleware(RequestDelegate next)
    {
        _next = next;
    }

    public async Task Invoke(HttpContext context, IMetricsCollector collector)
    {
        var sw = Stopwatch.StartNew();
        try
        {
            await _next(context);
        }
        finally
        {
            sw.Stop();
            // Fire-and-forget — we deliberately do not await to keep latency impact minimal
            _ = collector.RecordRequestAsync(
                path: context.Request.Path.Value ?? "/",
                method: context.Request.Method,
                statusCode: context.Response.StatusCode,
                latencyMs: sw.Elapsed.TotalMilliseconds);
        }
    }
}

File: slp/backend-nestjs/src/common/middleware/metrics.middleware.ts
import { Injectable, NestMiddleware, Inject } from '@nestjs/common';
import { Request, Response, NextFunction } from 'express';
import { METRICS_COLLECTOR } from '../../modules/metrics/metrics.module';
import type { IMetricsCollector } from '../../modules/metrics/metrics-collector.interface';

@Injectable()
export class MetricsMiddleware implements NestMiddleware {
  constructor(
    @Inject(METRICS_COLLECTOR) private metricsCollector: IMetricsCollector,
  ) {}

  use(req: Request, res: Response, next: NextFunction) {
    const start = Date.now();
    res.on('finish', () => {
      const latencyMs = Date.now() - start;
      this.metricsCollector.recordRequest(
        req.path,
        req.method,
        res.statusCode,
        latencyMs,
      );
    });
    next();
  }
}